api/types/types_drivers_executor.go

```diff
@@ -132,6 +158,7 @@ type ProvidesStorageExecutorCLI interface {
 // the driver implementations of the StorageExecutor interface.
 type StorageExecutorCLI interface {
 	StorageExecutorFunctions
+	StorageExecutorWithSupported
```

Comment:
Hi @codenrhoden,

Ah, good catch. It's the same reason I introduced the [`StorageDriverWithLogin`](https://github.com/emccode/libstorage/blame/master/api/types/types_drivers_storage.go#L160) interface in 6b6bd14fd3a9354951779be70af3620c909155af.

Let's look at how the `Login` function is [used](https://github.com/emccode/libstorage/blob/master/api/context/context.go#L127):

```go
// WithStorageSession returns a context that is logged into the storage
// platform.
func WithStorageSession(parent context.Context) (types.Context, error) {

	pctx := New(parent)

	d, ok := MustDriver(parent).(types.StorageDriverWithLogin)
	if !ok {
		pctx.Debug("driver is not StorageDriverWithLogin")
		return pctx, nil
	}

	sess, err := d.Login(pctx)
	if err != nil {
		pctx.WithError(err).Error("error logging into storage driver")
		return nil, err
	}
	if sess == nil {
		pctx.Warn("session is nil")
		return pctx, nil
	}

	return newContext(pctx, SessionKey, sess, nil, nil), nil
}
```

Please take note of this section of the above function:

```go
	d, ok := MustDriver(parent).(types.StorageDriverWithLogin)
	if !ok {
		pctx.Debug("driver is not StorageDriverWithLogin")
		return pctx, nil
	}
```

See how above I first check to see if the driver implements the interface `StorageDriverWithLogin` and I return early, not invoking the `Login` function if the provided driver does not have it?

This is a very clean manner in which Go makes it possible to make additive changes to code. Instead of adding the `Supported(Context, Store) (bool, error)` function to the existing `StorageExecutor` interface and breaking all the builds until that function is implemented for all the existing storage executor implementations, I simply created [a new interface](https://github.com/akutz/libstorage/blob/bugfix/missing-client-support/api/types/types_drivers_executor.go#L157-L173):

```go
// StorageExecutorWithSupported is an interface that executor implementations
// may use by defining the function "Supported(Context, Store) (bool, error)".
// This function indicates whether a storage platform is valid when executing
// the executor binary on a given client.
type StorageExecutorWithSupported interface {
	// Supported returns a flag indicating whether or not the platform
	// implementing the executor is valid for the host on which the executor
	// resides.
	Supported(
		ctx Context,
		opts Store) (bool, error)
}
```

So what does this mean? It means that the *existing* libStorage codebase continues to build and function as before without error. However, for those executors that *do* implement the interface I can [use that](https://github.com/akutz/libstorage/blob/bugfix/missing-client-support/cli/lsx/lsx.go#L71-L82):

```go
	if strings.EqualFold(cmd, apitypes.LSXCmdSupported) {
		op = "supported"
		if dws, ok := d.(apitypes.StorageExecutorWithSupported); ok {
			opResult, opErr := dws.Supported(ctx, store)
			if opErr != nil {
				err = opErr
			} else {
				result = opResult
			}
		} else {
			err = apitypes.ErrNotImplemented
		}
```

See above how we only invoke the `Supported` function if it is defined on that executor implementation? If it's not an `ErrNotImplemented` error is returned letting the eventual consumption of said error know that this executor simply hasn't yet implemented the new logic. And if it doesn't it will fall back to the first iteration of this PR where it's up to the individual executor functions to determine if something is supported.